### PR TITLE
Track and always resend both ConsoleColors

### PR DESCRIPTION
**Pull request by stephentoub**
As an alternative to https://github.com/dotnet/corefx/pull/4418, this commit tracks all changes made to foreground/background color through the APIs, and any time either foreground or background is updated, both are set.

### CODE REVIEW

**Comment by stephentoub** [comment]
cc: @weshaggard, @pallavit, @ellismg, @joshfree

**Comment by weshaggard** [comment]
I think this is a reasonable improvement over #4418 and I would be OK with this approach.

**Comment by ellismg** [comment]
I believe this is a workable approach.  I assume that we don't care about cases where someone outside of a managed application has changed one or more colors prior to starting the application or cases where someone is writing ANSI escape code directly into standard out to change things (the former seems like it could be likely, especially in cases where small applications orchestrate together to solve a larger problem, the latter seem fraught with peril since you'd have to reinvent a bunch of machinery like termcap processing, but might happen in cases where you're inter-oping with some existing native library).  Are there other cases where this falls down that I'm not thinking of?

My order of preference is to remove the API or continue to throw platform not supported exception, followed by doing this, but it's a much closer call.  I still feel strongly that #4418 isn't a workable solution.

**Comment by weshaggard** [comment]
Yes there are cases like you called out that will not work but I think that is probably true even outside the scope of these APIs, and I'm personally not super worried about those.

As for my preference I like this PR or something similar, I'm all for making as many code patterns work as possible. I don't prefer removing the getter's and we have discovered that throwing PSE causes too much pain so I don't believe either of those are real options.

**Comment by davidfowl** [comment]
This is a nice improvement

**Comment by pgavlin** [comment]
> I don't prefer removing the getter's and we have discovered that throwing PSE causes too much pain so I don't believe either of those are real options.

Do we have usage data for these getters?

**Comment by weshaggard** [comment]
>Do we have usage data for these getters?
We don't have any great usage data on this. I'm primarily basing my assumption on @stephentoub and @ellismg telling me that people keep hitting it while making Linux console apps, and thus prompting the question about potentially removing then in the first place.

**Comment by mellinoe** [comment]
What exactly do people use the Get* functions for? Personally, I've never used them so I'm curious what the actual use cases are. Presumably, you use them if different people are setting the color in different parts of code, and you want to know what the current state is (for whatever reason? I'm still not sure how you'd use that info). However, if some of those "different people" are in native code, then their effects are invisible, even with this change. You will get wrong values back. Am I wrong to assume that these methods are mainly useful for that, or do we think that most folks will just be using managed-only and won't be wrapping a native library, for example?

**Comment by stephentoub** [comment]
> What exactly do people use the Get* functions for? Personally, I've never used them so I'm curious what the actual use cases are

The most common case for the getters I've seen is the one I called out in #4418: you want to write out some text in a different color, so you get the current color, change the current color, write out some text, and set the current color back to whatever it was:
```C#
ConsoleColor origForeground = Console.ForegroundColor;
Console.ForegroundColor = ConsoleColor.Yellow;
Console.WriteLine("some text in yellow");
Console.ForegroundColor = origForeground;
```
Do a search across GitHub for "Console.ForegroundColor", and in the search result previews that come up, you'll see a lot of code that looks very similar to this snippet.

> You will get wrong values back...

This is one of the reasons several of us been pushing to remove them from the contracts, e.g. #4426.  As Matt outlined, the change in this PR is mostly reasonable, but it's also implementable by others on top of the set and reset members (after all those are the only operations we have under us on Unix), and it falls down in particular in the case you mention, where code elsewhere is mucking with the colors not through the Console APIs, e.g. by writing out ANSI escape sequences manually.

File: src/System.Console/src/System/ConsolePal.Unix.cs [hunk 1]
```diff
@@ -261,18 +261,42 @@ private static string GetCharset()
...
         {
-            // Get the numerical value of the color
-            int ccValue = (int)color;
-            if ((ccValue & ~0xF) != 0)
+            if (((int)value & ~0xF) != 0 && value != ConsoleColor.Unknown)
             {
                 throw new ArgumentException(SR.Arg_InvalidConsoleColor);
             }
 
+            lock (Console.Out)
```
Comment (bartonjs) [inline src/System.Console/src/System/ConsolePal.Unix.cs]: Are we not concerned about locking on something non-private?

File: src/System.Console/src/System/ConsolePal.Unix.cs [hunk 2]
```diff
@@ -306,6 +331,20 @@ private static void ChangeColor(bool foreground, ConsoleColor color)
             }
         }
 
+        /// <summary>Writes out the ANSI string to reset colors.</summary>
+        private static void WriteResetColorString()
+        {
+            // We only want to send the reset string if we're targeting a TTY device
+            if (!Console.IsOutputRedirected)
```
Comment (bartonjs) [inline src/System.Console/src/System/ConsolePal.Unix.cs]: Wouldn't we want this check in WriteSetColorString, as well?

Comment (stephentoub) [inline src/System.Console/src/System/ConsolePal.Unix.cs]: Yes.  It's there.

Comment (bartonjs) [inline src/System.Console/src/System/ConsolePal.Unix.cs]: Oops, didn't expand enough flyouts.

File: src/System.Console/src/System/ConsolePal.Unix.cs (hunk 1 above)
Comment (stephentoub) [inline src/System.Console/src/System/ConsolePal.Unix.cs]: We are.  But we previously changed this to lock on something private, and it ended up breaking scenarios, so we changed back as the lesser of two evils.  Customers lock on Console.Out in order to synchronize coloring with writing, so you can do things like:
```C#
lock (Console.Out)
{
    Console.ForegroundColor = ConsoleColor.Yellow;
    Console.WriteLine("yellow");
    Console.ForegroundColor = ConsoleColor.White;
}
```
and only "yellow" will be written in yellow, and not other text being written by other threads.  We have no other mechanism in the API for such synchronization, so we reverted. https://github.com/dotnet/corefx/pull/2855

**Comment by bartonjs** [comment]
As someone who has written the described pattern multiple times, I have to say that I like this approach better than removing it from the contracts.  If there are platforms/terminals/termcaps/etc that we can figure out how to get the originating color, we can always add that in place of the original Unknown.

**Comment by pallavit** [comment]
> As someone who has written the described pattern multiple times, I have to say that I like this approach better than removing it from the contracts. If there are platforms/terminals/termcaps/etc that we can figure out how to get the originating color, we can always add that in place of the original Unknown.

I agree. In fact, we can potentially follow the same approach with CursorVisible property as well.

**Comment by stephentoub** [comment]
> In fact, we can potentially follow the same approach with CursorVisible property as well.

CursorVisible returns a (non-nullable) Boolean... there are only two possible values :wink:.  We could track the previously set value and assume it hasn't changed via other means, but we can't know the state of the visibility when we entered, and we have no Unknown to represent that case without an API change.

**Comment by pallavit** [comment]
> We could track the previously set value and assume it hasn't changed via other means, but we can't know the state of the visibility when we entered.

I meant we can track previously set value only :).

**Comment by stephentoub** [comment]
> I meant we can track previously set value only :).

But then we may actually be lying the first time it's called. I don't love the ConsoleColor.Unknown solution, but at least there we're telling the truth, that the color is unknown.

**Comment by pallavit** [comment]
> But then we may actually be lying the first time it's called. I don't love the ConsoleColor.Unknown solution, but at least there we're telling the truth, that the color is unknown.

Fair point.

**Comment by stephentoub** [comment]
@dotnet-bot test this please (mac build timed out in package restore)

**Comment by ellismg** [comment]
@dotnet-bot test this please (#4396)

**Comment by ellismg** [comment]
LGTM.  The pragmatist in me feels like this is a good enough compromise.  You can certainly construct cases where it will do the wrong thing, but for a vast majority of apps, this will be OK, I believe.  Given that we want to take a stance that we try not to fracture existing types apart and bias towards including more surface area, doing this feels like the right move.

**Comment by stephentoub** [comment]
Agreed.  Thanks, Matt.

**Comment by weshaggard** [comment]
Thanks @stephentoub

**Pull request merged**